org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/ServiceRule.java

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright (c) OSGi Alliance (2019). All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.osgi.test.junit4;
+
+import static org.osgi.test.common.osgi.Filters.format;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.SortedMap;
+import java.util.stream.Collectors;
+
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+import org.osgi.framework.Filter;
+import org.osgi.framework.ServiceReference;
+import org.osgi.test.common.osgi.ServiceHolder;
+
+/**
+ * A JUnit 4 Rule to depend on OSGi services.
+ * <p>
+ * Example:
+ * <br>
+ * <pre>
+ * {@literal @}Rule
+ * public ServiceRule&lt;Foo> foos = new ServiceRule.Builder&lt;>(Foo.class)
+ *     .filter("(service.description=Acme Inc.)")
+ *     .requiredWithin(300)
+ *     .build();
+ * </pre>
+ *
+ * @param <T> the service type
+ */
+public class ServiceRule<T> extends BundleContextRule implements TestRule {
```

Comment:
How about `ServiceUseRule` since it more accurately describes the purpose as distinct from a `ServiceRegistrationRule`.

Reply:
We should take care not to subclass. Better to aggregate since that allows us more freedom to refactor. Cannot one rule depend upon another rule?

Reply:
I do not believe there is a interdependency mechanism for extensions. but I'll look.

Reply:
ok, so there is certainly no such mechanism in junit4, maybe in Junit5.

Reply:
the issue with not using inheritance is that making shared logic that we'll use in both junit4+5 will be hard, it will require pass through delegation (`rule.delegate.doStuff()`) which will make every use uglier than the previous.

Reply:
But I'm open to suggestions.

Reply:
what about straight up composition:
```
	@Rule
	public BundleContextRule	bundleContextRule	= new BundleContextRule();
	@Rule
	public ServiceUseRule<Foo>	fooUse	= new ServiceUseRule.Builder<>(Foo.class, bundleContextRule).build();

	@Test
	public void test() throws Exception {
		assertThat(fooUse.getService()).isNotNull();
	}
```

Reply:
this type of inheritance cannot work in junit4 since there's no way to control the order in which the rules are initialized, and as is the case with these two rules, the BundleContextRule must initialize before the ServiceUseRule.

Reply:
k, it can be made to work.

Reply:
I thought that JUnit 4 always initializes rules in the order that they are declared, so composition should be fine as long as you declare the BundleContextRule first. But I can't find any docs on that.

There's also the RuleChain rule: https://stackoverflow.com/questions/2730365/how-to-handle-ordering-of-rules-when-they-are-dependent-on-each-other